validator_client/src/beacon_node_fallback.rs

```diff
@@ -475,225 +547,88 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
 
     /// Run `func` against each candidate in `self`, returning immediately if a result is found.
     /// Otherwise, return all the errors encountered along the way.
-    ///
-    /// First this function will try all nodes with a suitable status. If no candidates are suitable
-    /// or all the requests fail, it will try updating the status of all unsuitable nodes and
-    /// re-running `func` again.
-    pub async fn first_success<'a, F, O, Err, R>(
-        &'a self,
-        require_synced: RequireSynced,
-        offline_on_failure: OfflineOnFailure,
-        func: F,
-    ) -> Result<O, Errors<Err>>
+    pub async fn first_success<F, O, Err, R>(&self, func: F) -> Result<O, Errors<Err>>
     where
-        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        F: Fn(BeaconNodeHttpClient) -> R,
         R: Future<Output = Result<O, Err>>,
         Err: Debug,
     {
+        // First pass: try `func` on all candidates. Candidate order has already been set in
+        // `update_all_candidates`. This ensures the most suitable node is always tried first.
+        let candidates = self.candidates.read().await;
```

Comment:
This read lock is held for the duration until the calls completes, which could last for a few seconds depend on how fast the API returns I think?